### ISSUE TITLE
Add EAN8 or EAN13 barcode generation in case none is available

### ISSUE THREAD

**Issue by daN4cat** [jekkos/opensourcepos#111]
I have already this feature working in my old fork that I could add to master:
 in case the field is empty and the user selects the checkbox and option Generate barcode, a EAN8 barcode is generated using item_id.
The barcode label can then be used to mark products and it's unique for the system since item_id is progressive.

If no objections I would proceed in implementing this feature.

**Comment by jekkos** [comment]
If I'm not mistaken then there should already be something in place to handle this situation. In 2.3.2 you can select a preferred method for barcode generation in the config screen (item_id or upc number). If you select upc and there's no value, then code will fall back to the item_id and generate the barcode using code39 or code128

Are there any other advantages in using EAN-8? Would be nice to have it implemented in a similar way as the other generators now, depending on the amount of work involved.

**Comment by jekkos** [comment]
Maybe the item_id used for barcode generation should be filled in in the upc field as well then? Any suggestions are welcome.

**Comment by daN4cat** [comment]
OK, I missed something there. 
For sure the generated barcode should go in the upc field because if you are sync'ing items with an external system you may want to use that field.
But let me play a bit more with what's implemented already and come back.

**Comment by daN4cat** [comment]
I now remember why EAN8. In my wife's shop they use a Casio label printer which natively supports printing EAN8 and EAN13 barcodes. I used EAN8 because it's shorter but I see a point in having that barcode support.

**Comment by daN4cat** [comment]
You can assign this issue to me if you are fine.

**Comment by daN4cat** [comment]
An example of Label printer with native EAN8 & EAN13 support: https://www.casio.co.uk/products/label-products/Product/KL-7400/

**Comment by daN4cat** [comment]
I quickly checked the current barcode generation classes and they come from a github that has only js code now. I wonder where those php files were coming from.
It will take some time to reuse the same class hierarchy to generate EAN codes because of the maps and etc.

**Comment by daN4cat** [comment]
I could find from the github the EAN13 js code, not EAN8.
I think I will proceed with EAN13 and discard EAN8. In the end EAN13 is supported by Casio label printer.

**Comment by daN4cat** [comment]
OK, I implemented the EAN8 and EAN13  barcode generators using the same base class and structure as for Code39 and Code128. The selection is done from the same config page.

I have still to sort out an issue about setting the correct pix per bar or centre the barcode in the image, either will do.

However what I really need at the end of this is to set a code in the UPC field based on the item_id, if empty.

I think we should do only if empty to avoid potential mistakes where people might select a valid one and override it. Or do we prefer to override it always?

Any view?

**Comment by jekkos** [comment]
Currently there's a config option for this feature already, I think the default behavior is to use the upc field, and if empty, use the item_id instead for barcode generation.

Once this is done there's still the possibility to add a new upc code after the generation. But this shouldn't break search functionality, as it's well capable of searching for upc codes or item_ids.

**Comment by daN4cat** [comment]
OK, I meant something else. The UPC field is not filled in if empty and you want to use the item_id for the barcode generation.

Nothing to do with search or else.

**Comment by daN4cat** [comment]
OK, I got better what you meant with if "UPC is empty" the item_id is used. I saw that in the code and I fixed an issue in another line around that point.

I'm not sure I need to fill the UPC field (or item_number) in case of Code 39 and Code 128, for sure I need to do for EAN8 & EAN13, because those are newly generated barcodes seeded with item_id.

Any view whether I should always set UPC field if empty when a code is generated, or just avoid for Code39/128?

**Comment by jekkos** [comment]
If I'm not mistaken then there is a setting in the barcode config that defines which field to use for the generatoin. If it's set to upc which can be empty for some items, then fallback will be done to item_id, regardless of which barcode generation algorithm is used.

I didn't check the code this time but I think that the fallback should be kept outside of the specific algorithm's generation implementation.

In this case I would just leave the upc field empty so that it can still be overridden afterwards. Generated barcodes based on item_id will never invalidate as the register search will work with either item_id or item_number..

**Comment by daN4cat** [comment]
Yes, the fallback is in place.

For EAN you cannot leave the UPC field empty because it's a different code using item_id as seed.

So, I need to leave empty only for Code39/128 as per your suggestion.

Also I need to add a checkbox in Items to search for only empty UPC because at the moment they go bottom of the list. It used to be possible to sort the table by UPC and get the empty one top of the table. I don't know why all the table sorting of the past have been removed.

**Comment by daN4cat** [comment]
Code submitted for review and integration if all fine.

**Comment by jekkos** [comment]
Had a look at the code, pulled it in and now integrated in master. :+1:

**Comment by daN4cat** [comment]
oh great, it could have been faster I must admit :-)

I feel better after the half fiasco of the Takings story :-(

**Comment by jekkos** [comment]
Yes first time right. Thanks for the contrib!

**Comment by glassen** [comment]
I think I've found somewhat of a bug in this feature. I'm starting from scratch with a completely new install. Pardon me if this isn't the correct way to comment on the feature on a closed issue.

1. In barcode setup i choose EAN-13 since it's what's used on our barcode labeled items.
2. Add items without EAN-13 barcodes since they are either not physical items or at least not labeled. I do not add any EAN number.
3. Select the items without EAN and choose generate barcodes.

The first time I do this I get beautiful barcodes with correct data, but if I select the same barcodes again, now that they have an automatically generated EAN, I get another barcode that doesn't decode correctly.

![correct](https://cloud.githubusercontent.com/assets/14387561/10461648/925863f6-71da-11e5-9650-bb19de6d3174.png)
Correct barcode generated the first time without EAN filled

![error](https://cloud.githubusercontent.com/assets/14387561/10461655/9d71631e-71da-11e5-914f-33224bf77c48.png)
Erroneous barcode generated when EAN is filled

Is this reproducible for anyone else?

**Comment by daN4cat** [comment]
Hi,

I've checked with Barcode Scanner (Android application) and the first is a EAN 13 (2001000000005)), the second is UPC_E (14075847).

I tested with my software (config is EAN13 and Barcode Content is Item Id/Name):

- empty barcode item
- click on generate barcode
- get EAN13 barcode based on Item Id
- select the item and that shows a EAN13 code (200....)
- click generate barcode
- get EAN13 barcode picture as previous one

Here it's all fine.

However if I select in config Barcode Content UPC/EAN/ISBN and still EAN13 I don't get the barcode as before, because it tries to generate a barcode code out of a different seed (it was Item Id before).

Not sure what we to do here if we change seed. I need to think about.

**Comment by glassen** [comment]
I'm sorry, it seems that I've missed that crucial detail in my description. I have selected it to be UPC/EAN/ISBN as Barcode Content from the beginning.

**Comment by daN4cat** [comment]
OK I need to check that case.

**Comment by daN4cat** [comment]
OK, there is an issue here. Even if I fixed the generation of barcodes avoiding to create new ones out of exiting I cannot with the current code avoid printing a UPC as EAN13. But that might be fine, the issue is that you cannot print a UPC as a EAN8.

I suppose that the ideal behaviour would be to print the barcodes according to the format, so if they are EANs we print in the correct EAN format and the rest as Code39/128.
However this requires decoding of the code before selecting which format. And that's something I don't have in terms of algorithm at the moment.

Alternative is to generate EANs if required (different config item) and print in Code39/128. That should work always. If an item has an exiting barcode in EAN on the box that will work anyway because correctly scanned.

**Comment by daN4cat** [comment]
This is now fixed.
